ext/mbstring/libmbfl/filters/mbfilter_uhc.c

```diff
@@ -196,12 +196,16 @@ static size_t mb_uhc_to_wchar(unsigned char **in, size_t *in_len, uint32_t *buf,
 	unsigned char *p = *in, *e = p + *in_len;
 	uint32_t *out = buf, *limit = buf + bufsize;
 
+	e--; /* Stop the main loop 1 byte short of the end of the input */
+
```

Comment:
Wouldn't this also be solved by reducing the ``limit`` by 1?

Reply:
No. The point of this is that while we check `p < e` once on every iteration of the main `while` loop, we also have to check it a *second* time *inside* the `while` loop, before we pull the 2nd byte of what looks like a 2-byte character.

By never running the main `while` loop to the very last byte of the input string, it means we don't have to do this extra bounds check *inside* the `while` loop.

Reply:
By the way, it would be even better if we didn't have to do any `p < e` bounds checks *at all*... and if we can assume that every input string is always null-terminated, we could use the null terminator as a sentinel to achieve that (as I mentioned above).

I just really doubt whether that's a safe assumption to make when the PHP lexer calls into mbstring. For functions like `mb_convert_encoding`, it's absolutely a safe assumption.

Reply:
Ah right okay.

Somehow, I would expect the lexer to actually null terminate strings. Maybe add an assertion to check the string is null terminated, and then try to come up with a case that makes this assertion fail? If you find such a case then having it in the test suite is good, otherwise I think we can assume they are always null terminated.

Reply:
OK, I'll give that a try.

If it works, then a lot of these functions can be made faster still, while (perhaps) losing less readability.

Reply:
I pretty sure that the lexer does not NUL terminate; otherwise there wouldn't be the need for the `yytext`/`yyleng` pairs.

Reply:
@cmb69 You are absolutely right; adding an assertion and running the test suite just proved it.

So much for that idea.

Reply:
boo, well that sucks :(

Reply:
Doesn't it, though? Especially since `zend_string`s *are* null-terminated, grrr...